core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushMergeWriterUpdateIntoConnector.java

```diff
@@ -0,0 +1,157 @@
+        this.typeAnalyzer = requireNonNull(typeAnalyzer, "typeAnalyzer is null");
+        this.plannerContext = requireNonNull(plannerContext, "plannerContext is null");
+    }
+
+    private final Metadata metadata;
+    private final PlannerContext plannerContext;
+    private final TypeAnalyzer typeAnalyzer;
+
+    @Override
+    public Pattern<TableFinishNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(TableFinishNode node, Captures captures, Context context)
+    {
+        TableScanNode tableScan = captures.get(TABLE_SCAN);
+        ProjectNode project = captures.get(PROJECT_NODE_CAPTURE);
+        MergeWriterNode mergeWriter = captures.get(MERGE_WRITER_NODE_CAPTURE);
+
+        List<String> orderedColumnNames = mergeWriter.getTarget().getMergeParadigmAndTypes().getColumnNames();
+        List<? extends Node> mergeAssignments = project.getAssignments().get(new Symbol("merge_row")).getChildren();
+
+        if (mergeAssignments.size() < orderedColumnNames.size()) {
+            return Result.empty();
+        }
+        ImmutableMap.Builder<String, Domain> assignmentsBuilder = ImmutableMap.builder();
+
+        for (int i = 0; i < orderedColumnNames.size(); i++) {
+            String columnName = orderedColumnNames.get(i);
+            Node assigmentNode = mergeAssignments.get(i);
+            checkForFailNode(assigmentNode);
+
+            ConnectorExpressionTranslator.ConnectorExpressionTranslation expressionTranslation = ConnectorExpressionTranslator.translateConjuncts(
+                    context.getSession(),
+                    ((Expression) assigmentNode),
+                    context.getSymbolAllocator().getTypes(),
+                    plannerContext,
+                    typeAnalyzer);
+
+            ConnectorExpression connectorExpression = expressionTranslation.connectorExpression();
+
+            if (connectorExpression instanceof Constant constantExpression) {
+                assignmentsBuilder.put(columnName, Domain.singleValue(constantExpression.getType(), constantExpression.getValue()));
+            }
+        }
+        Map<String, Domain> assignments = assignmentsBuilder.buildOrThrow();
+        if (assignments.isEmpty()) {
+            return Result.empty();
+        }
+
+        return metadata.applyUpdate(context.getSession(), tableScan.getTable())
+                .map(newHandle -> new TableUpdateNode(
+                        context.getIdAllocator().getNextId(),
+                        newHandle,
+                        assignments,
+                        getOnlyElement(node.getOutputSymbols())))
+                .map(Result::ofPlanNode)
+                .orElseGet(Result::empty);
+    }
+
+    private static void checkForFailNode(Node node)
+    {
+        if (node instanceof Cast castNode) {
+            Expression expression = castNode.getExpression();
+            if (expression.toString().startsWith("\"@fail")) {
```

Comment:
seems fragile, is there a better way to test that? What if we push a failed node?

Reply:
you are asserting on the serialized function representation. this could potentially change when `ResolvedFunction` implementation has changed. You could use `MetadataManager.decodeFunction` instead

Reply:
https://github.com/trinodb/trino/pull/16445#discussion_r1169712863